File: sdk-java/src/main/java/io/littlehorse/sdk/wfsdk/Workflow.java
package io.littlehorse.sdk.wfsdk;

import com.google.protobuf.InvalidProtocolBufferException;
import com.google.protobuf.Message;
import com.google.protobuf.util.JsonFormat;
import io.grpc.Status.Code;
import io.grpc.StatusRuntimeException;
import io.littlehorse.sdk.common.LHLibUtil;
import io.littlehorse.sdk.common.proto.AllowedUpdateType;
import io.littlehorse.sdk.common.proto.ExponentialBackoffRetryPolicy;
import io.littlehorse.sdk.common.proto.GetLatestWfSpecRequest;
import io.littlehorse.sdk.common.proto.LittleHorseGrpc.LittleHorseBlockingStub;
import io.littlehorse.sdk.common.proto.PutExternalEventDefRequest;
import io.littlehorse.sdk.common.proto.PutWorkflowEventDefRequest;
import io.littlehorse.sdk.common.proto.PutWfSpecRequest;
import io.littlehorse.sdk.common.proto.ThreadRetentionPolicy;
import io.littlehorse.sdk.common.proto.WfSpecId;
import io.littlehorse.sdk.common.proto.WorkflowRetentionPolicy;
import io.littlehorse.sdk.wfsdk.internal.ExternalEventNodeOutputImpl;
import io.littlehorse.sdk.wfsdk.internal.ThrowEventNodeOutputImpl;
import io.littlehorse.sdk.wfsdk.internal.WorkflowImpl;
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Set;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.tuple.Pair;

/** The Workflow class represents a `WfSpec` object in the API. */
@Slf4j
public abstract class Workflow {

    protected ThreadFunc entrypointThread;
    protected String name;
    protected PutWfSpecRequest.Builder spec;
    protected Queue<Pair<String, ThreadFunc>> threadFuncs;
    protected Integer defaultTaskTimeout;
    protected WorkflowRetentionPolicy wfRetentionPolicy;
    protected ThreadRetentionPolicy defaultThreadRetentionPolicy;
    protected String parentWfSpecName;

    protected ExponentialBackoffRetryPolicy defaultExponentialBackoff;
    protected int defaultSimpleRetries;
    protected Set<ExternalEventNodeOutputImpl> externalEventsToRegister;
    protected Set<ThrowEventNodeOutputImpl> workflowEventsToRegister;

    /**
     * Internal constructor used by WorkflowImpl.
     *
     * @param name name of `WfSpec`.
     * @param entrypointThreadFunc is the entrypoint thread function.
     */
    protected Workflow(String name, ThreadFunc entrypointThreadFunc) {
        this.threadFuncs = new LinkedList<>();
        this.entrypointThread = entrypointThreadFunc;
        this.name = name;
        this.spec = PutWfSpecRequest.newBuilder().setName(name);
        this.externalEventsToRegister = new HashSet<>();
        this.workflowEventsToRegister = new HashSet<>();
    }

    /**
     * Makes this a child workflow.
     * @param parentWfSpecName is the name of the parent wfSpec.
     */
    public void setParent(String parentWfSpecName) {
        this.parentWfSpecName = parentWfSpecName;
    }

    /**
     * Sets the default timeout for all TaskRun's in this workflow.
     * @param timeoutSeconds is the value for the timeout to set.
     */
    public void setDefaultTaskTimeout(int timeoutSeconds) {
        this.defaultTaskTimeout = timeoutSeconds;
    }

    /**
     * Returns the default task timeout, or null if it is not set.
     * @return the default task timeout for this Workflow.
     */
    public Integer getDefaultTaskTimeout() {
        return defaultTaskTimeout;
    }

    /**
     * Tells the Workflow to configure (by default) a Simple Retry Policy for every Task Node. Passing
     * a value of '1' means that there will be one retry upon failure. Retries are scheduled immediately
     * without delay.
     *
     * Can be overriden by setting the retry policy on the WorkflowThread or TaskNodeOutput level.
     * @param defaultSimpleRetries is the number ofretries to attempt.
     */
    public void setDefaultTaskRetries(int defaultSimpleRetries) {
        if (defaultSimpleRetries < 0) {
            throw new IllegalArgumentException("Cannot have negative retries!");
        }
        this.defaultSimpleRetries = defaultSimpleRetries;
    }

    /**
     * Tells the Workflow to configure (by default) the specified ExponentialBackoffRetryPolicy as
     * the retry policy.
     *
     * Can be overriden by setting the retry policy on the WorkflowThread or TaskNodeOutput level.
     * @param defaultPolicy is the Exponential Backoff Retry Policy to configure by default for all
     * Task Nodes.
     */
    public void setDefaultTaskExponentialBackoffPolicy(ExponentialBackoffRetryPolicy defaultPolicy) {
        this.defaultExponentialBackoff = defaultPolicy;
    }

    /**
     * Gets the workflow name passed at {@link #newWorkflow(String, ThreadFunc)}
     *
     * @return the Workflow name
     */
    public String getName() {
        return name;
    }

    /**
     * Sets the retention policy for all WfRun's created by this WfSpec.
     * @param policy is the Workflow Retention Policy.
     * @return this Workflow.
     */
    public Workflow withRetentionPolicy(WorkflowRetentionPolicy policy) {
        this.wfRetentionPolicy = policy;
        return this;
    }

    /**
     * Sets the retention policy for all ThreadRun's belong to this WfSpec.
     *
     * Note that each Thread can override the configured Retention Policy by
     * using WorkflowThread#withRetentionPolicy.
     * @param policy is the Workflow Retention Policy.
     * @return this Workflow.
     */
    public Workflow withDefaultThreadRetentionPolicy(ThreadRetentionPolicy policy) {
        this.defaultThreadRetentionPolicy = policy;
        return this;
    }

    /**
     * Creates a new Workflow with the provided name and entrypoint thread function.
     *
     * @param name is the name of the `WfSpec`.
     * @param entrypointThreadFunc is the ThreadFunc for the entrypoint ThreadSpec.
     * @return a Workflow.
     */
    public static Workflow newWorkflow(String name, ThreadFunc entrypointThreadFunc) {
        return new WorkflowImpl(name, entrypointThreadFunc);
    }

    /**
     * Defines the type of update to perform when saving the WfSpec:
     * AllowedUpdateType.ALL (Default): Creates a new WfSpec with a different version (either major or revision).
     * AllowedUpdateType.MINOR_REVISION_ONLY: Creates a new WfSpec with a different revision if the change is a major version it fails.
     * AllowedUpdateType.NONE: Fail with the ALREADY_EXISTS response code.
     * @param allowedUpdateType
     * @return this Worflow
     */
    public Workflow withUpdateType(AllowedUpdateType allowedUpdateType) {
        this.spec.setAllowedUpdates(allowedUpdateType);
        return this;
    }

    /**
     * Compiles this Workflow into a `WfSpec`.
     *
     * @return a `PutWfSpecRequest` that can be used for the gRPC putWfSpec() call.
     */
    public abstract PutWfSpecRequest compileWorkflow();

    /**
     * Returns the names of all `TaskDef`s used by this workflow.
     *
     * @return a Set of Strings containing the names of all `TaskDef`s used by this workflow.
     */
    public abstract Set<String> getRequiredTaskDefNames();

    /**
     * Returns the names of all `ExternalEventDef`s used by this workflow. Includes
     * ExternalEventDefs used for Interrupts or for EXTERNAL_EVENT nodes.
     *
     * @return a Set of Strings containing the names of all `ExternalEventDef`s used by this
     *     workflow.
     */
    public abstract Set<String> getRequiredExternalEventDefNames();

    /**
     * Returns ExternalEventDef registrations declared via ExternalEventNodeOutput#registeredAs.
     *
     * @return a Set of PutExternalEventDefRequest for ExternalEventDefs that will be registered
     *     with this workflow.
     */
    public abstract Set<PutExternalEventDefRequest> getExternalEventDefsToRegister();

    /**
     * Returns the names of all `WorkflowEventDef`s used by this workflow.
     *
     * @return a Set of Strings containing the names of all `WorkflowEventDef`s thrown by this
     *      workflow.
     */
    public abstract Set<String> getRequiredWorkflowEventDefNames();

    /**
     * Returns WorkflowEventDef registrations declared via ThrowEventNodeOutput#registeredAs.
     *
     * @return a Set of PutWorkflowEventDefRequest for WorkflowEventDefs that will be registered
     *     with this workflow.
     */
    public abstract Set<PutWorkflowEventDefRequest> getWorkflowEventDefsToRegister();

    /**
     * Returns the associated PutWfSpecRequest in JSON form.
     *
     * @return the associated PutWfSpecRequest in JSON form.
     */
    public String compileWfToJson() {
        PutWfSpecRequest wfSpec = compileWorkflow();
        try {
            return JsonFormat.printer().includingDefaultValueFields().print(wfSpec);
        } catch (InvalidProtocolBufferException exn) {
            throw new RuntimeException(exn);
        }
    }

    /**
     * Checks if the WfSpec exists
     *
     * @param client is an LHClient.
     * @return true if the workflow spec is registered or false otherwise
     */
    public boolean doesWfSpecExist(LittleHorseBlockingStub client) {
        try {
            client.getLatestWfSpec(
                    GetLatestWfSpecRequest.newBuilder().setName(name).build());
            return true;
        } catch (StatusRuntimeException exn) {
            if (exn.getStatus().getCode() == Code.NOT_FOUND) {
                return false;
            }

            throw exn;
        }
    }

    /**
     * Checks if the WfSpec exists for a given version
     *
     * @param client is an LHClient.
     * @param majorVersion the workflow Major Version
     * @return true if the workflow spec is registered for this Major Version or false otherwise
     */
    public boolean doesWfSpecExist(LittleHorseBlockingStub client, Integer majorVersion) {
        // TODO: LH-282, support revision versioning here.
        if (majorVersion == null) return doesWfSpecExist(client);

        try {
            client.getWfSpec(WfSpecId.newBuilder()
                    .setName(name)
                    .setMajorVersion(majorVersion)
                    .build());
            return true;
        } catch (StatusRuntimeException exn) {
            if (exn.getStatus().getCode() == Code.NOT_FOUND) {
                return false;
            }

            throw exn;
        }
    }

    /**
     * Deploys the WfSpec object to the LH Server. Registering the WfSpec via
     * Workflow::registerWfSpec() is the same as client.putWfSpec(workflow.compileWorkflow()).
     *
     * @param client is an LHClient.
     */
    public abstract void registerWfSpec(LittleHorseBlockingStub client);

    /**
     * Writes out the PutWfSpecRequest in JSON form in a directory.
     *
     * @param directory is the location to save the resources.
     */
    public void compileAndSaveToDisk(String directory) {
        PutWfSpecRequest wf = compileWorkflow();
        String wfFileName = wf.getName() + "-wfspec.json";
        log.info("Saving WfSpec to {}", wfFileName);
        saveProtoToFile(directory, wfFileName, wf);
    }

    private static void saveProtoToFile(String directory, String fileName, Message content) {
        Path path = Paths.get(directory, fileName);
        try {
            File file = new File(path.toString());
            file.getParentFile().mkdirs();
            Files.write(path, LHLibUtil.protoToJson(content).getBytes());
        } catch (IOException exn) {
            throw new RuntimeException(exn);
        }
    }
}

File: sdk-dotnet/LittleHorse.Sdk/Workflow/Spec/Workflow.cs
using Google.Protobuf;
using Grpc.Core;
using LittleHorse.Sdk.Common.Proto;
using LittleHorse.Sdk.Helper;
using Microsoft.Extensions.Logging;

namespace LittleHorse.Sdk.Workflow.Spec;

using static LittleHorse.Sdk.Common.Proto.LittleHorse;

/// <summary>
/// The <c>Workflow</c> class represents a workflow definition in the LittleHorse system.
/// </summary>
public class Workflow
{
    private readonly ILogger<Workflow>? _logger;
    private string _name;
    private readonly Action<WorkflowThread> _entryPoint;
    private PutWfSpecRequest? _compiledWorkflow;
    private readonly PutWfSpecRequest _spec;
    private readonly Queue<Tuple<string, Action<WorkflowThread>>> _threadActions;
    private readonly string _parentWfSpecName;
    private readonly HashSet<string> _requiredTaskDefNames;
    private readonly HashSet<string> _requiredEedNames;
    private readonly HashSet<string> _requiredWorkflowEventDefNames;
    private int _defaultTaskTimeout;
    private int _defaultSimpleRetries;
    private ExponentialBackoffRetryPolicy _defaultExponentialBackoff = null!;
    private ThreadRetentionPolicy? _defaultThreadRetentionPolicy;
    private WorkflowRetentionPolicy? _wfRetentionPolicy;
    internal readonly Stack<WorkflowThread> Threads;
    private readonly List<ThrowEventNodeOutput> _workflowEventsToRegister = new();

    /// <summary>
    /// Initializes a new instance of the <see cref="Workflow"/> class.
    /// </summary>
    /// <param name="name">The name of the workflow.</param>
    /// <param name="entryPoint">This is the main entrypoint for the workflow logic.</param>
    public Workflow(string name, Action<WorkflowThread> entryPoint)
    {
        _logger = LHLoggerFactoryProvider.GetLogger<Workflow>();
        _name = name;
        _parentWfSpecName = string.Empty;
        _entryPoint = entryPoint;
        _compiledWorkflow = null!;
        _spec = new PutWfSpecRequest { Name = name };
        _threadActions = new Queue<Tuple<string, Action<WorkflowThread>>>();
        _requiredTaskDefNames = new HashSet<string>();
        _requiredEedNames = new HashSet<string>();
        _requiredWorkflowEventDefNames = new HashSet<string>();
        Threads = new Stack<WorkflowThread>();
        // Force workflow construction here so all event registrations are done before Compile
        // var wfThread = new WorkflowThread(this, entryPoint);
        // _spec.ThreadSpecs.Add("entrypoint", wfThread.Compile());
    }
    
    /// <summary>
    /// Compiles this Workflow into a `WfSpec`.
    /// </summary>
    /// <returns>
    /// A `PutWfSpecRequest` that can be used for the gRPC putWfSpec() call.
    /// </returns>
    public PutWfSpecRequest Compile()
    {
        return _compiledWorkflow ??= CompileWorkflowDetails();
    }

    /// <summary>
    /// Deploys the WfSpec object to the LH Server. Registering the WfSpec via
    /// Workflow::RegisterWfSpec() is the same as client.putWfSpec(workflow.compileWorkflow()).
    /// </summary>
    /// <param name="client">
    /// It is a LHClient.
    /// </param>
    public void RegisterWfSpec(LittleHorseClient client)
    {
        var request = Compile();
        foreach (var node in _workflowEventsToRegister)
        {
            client.PutWorkflowEventDef(node.ToPutWorkflowEventDefRequest());
            _logger!.LogInformation($"Registered WorkflowEventDef: {node.ToPutWorkflowEventDefRequest().Name}");
        }
        _logger!.LogInformation($"Created wfSpec:\n{LHMappingHelper.ProtoToJson(client.PutWfSpec(request))}");

    }

    internal string AddSubThread(string subThreadName, Action<WorkflowThread> subThreadAction) 
    {
        foreach (var threadPair in _threadActions)
        {
            if (threadPair.Item1 == subThreadName)
            {
                throw new Exception($"Thread {subThreadName} already exists");
            }
        }
        
        _threadActions.Enqueue(new Tuple<string, Action<WorkflowThread>>(subThreadName, subThreadAction));
        return subThreadName;
    }
    
    private PutWfSpecRequest CompileWorkflowDetails( ) 
    {
        string entrypointThreadName = AddSubThread("entrypoint", _entryPoint);
        _spec.EntrypointThreadName = entrypointThreadName;

        while (_threadActions.Count != 0) 
        {
            Tuple<string, Action<WorkflowThread>> nextThreadAction = _threadActions.Dequeue();
            string actionName = nextThreadAction.Item1;
            Action<WorkflowThread> threadAction = nextThreadAction.Item2;
            var wfThread = new WorkflowThread(this, threadAction);
            _spec.ThreadSpecs.Add(actionName, wfThread.Compile());
        }
        
        if (_wfRetentionPolicy != null) 
        {
            _spec.RetentionPolicy = _wfRetentionPolicy;
        }
        
        if (!string.IsNullOrEmpty(_parentWfSpecName))
        {
            _spec.ParentWfSpec = new WfSpec.Types.ParentWfSpecReference { WfSpecName = _parentWfSpecName };
        }

        return _spec;
    }
    
    internal void AddTaskDefName(string taskDefName) 
    {
        _requiredTaskDefNames.Add(taskDefName);
    }
    
    internal void AddExternalEventDefName(string eedName) 
    {
        _requiredEedNames.Add(eedName);
    }
    
    /// <summary>
    /// Returns the names of all `TaskDef`s used by this workflow.
    /// </summary>
    /// <returns>
    /// A HashSet of strings containing the names of all `TaskDef`s used by this workflow.
    /// </returns>
    public HashSet<string> GetRequiredTaskDefNames()
    {
        return _requiredTaskDefNames;
    }

    /// <summary>
    /// Returns the default task timeout, or null if it is not set.
    /// </summary>
    /// <returns>
    /// the default task timeout for this Workflow.
    /// </returns>
    public int GetDefaultTaskTimeout() 
    {
        return _defaultTaskTimeout;
    }

    /// <summary>
    /// Sets the default timeout for all TaskRun's in this workflow.
    /// </summary>
    /// <param name="timeoutSeconds">
    /// It is the value for the timeout to set.
    /// </param>
    public void SetDefaultTaskTimeout(int timeoutSeconds) 
    {
        _defaultTaskTimeout = timeoutSeconds;
    }
    
    internal int GetDefaultSimpleRetries() 
    {
        return _defaultSimpleRetries;
    }
    
    /// <summary>
    /// Tells the Workflow to configure (by default) a Simple Retry Policy for every Task Node. Passing
    /// a value of '1' means that there will be one retry upon failure. Retries are scheduled immediately
    /// without delay.
    ///
    /// Can be overriden by setting the retry policy on the WorkflowThread or TaskNodeOutput level.
    /// </summary>
    /// <param name="defaultSimpleRetries">
    /// It is the number of retries to attempt.
    /// </param>
    public void SetDefaultTaskRetries(int defaultSimpleRetries) 
    {
        if (defaultSimpleRetries < 0) 
        {
            throw new ArgumentException("Cannot have negative retries!");
        }
        _defaultSimpleRetries = defaultSimpleRetries;
    }
    
    internal ExponentialBackoffRetryPolicy? GetDefaultExponentialBackoffRetryPolicy() 
    {
        return _defaultExponentialBackoff!;
    }
    
    /// <summary>
    /// Tells the Workflow to configure (by default) the specified ExponentialBackoffRetryPolicy as
    /// the retry policy.
    ///
    /// Can be overriden by setting the retry policy on the WorkflowThread or TaskNodeOutput level.
    /// </summary>
    /// <param name="defaultPolicy">
    /// It is the Exponential Backoff Retry Policy to configure by default for all Task Nodes.
    /// </param>
    public void SetDefaultTaskExponentialBackoffPolicy(ExponentialBackoffRetryPolicy defaultPolicy) 
    {
        _defaultExponentialBackoff = defaultPolicy;
    }

    /// <summary>
    /// Gets the workflow name passed at <code>new Workflow(string, Action)</code>
    /// </summary>
    /// <returns>
    /// The Workflow name
    /// </returns>
    public string GetName() 
    {
        return _name;
    }
    
    /// <summary>
    /// Returns the names of all `ExternalEventDef`s used by this workflow. Includes
    /// ExternalEventDefs used for Interrupts or for EXTERNAL_EVENT nodes.
    /// 
    /// </summary>
    /// <returns>
    /// A HashSet of strings containing the names of all `ExternalEventDef`s used by this workflow.
    /// </returns>
    public HashSet<string> GetRequiredExternalEventDefNames()
    {
        _compiledWorkflow ??= CompileWorkflowDetails();
        return _requiredEedNames;
    }
    
    /// <summary>
    /// Returns the names of all `WorkflowEventDef`s used by this workflow.
    /// 
    /// </summary>
    /// <returns>
    /// A HashSet of strings containing the names of all `WorkflowEventDef`s thrown by this workflow.
    /// </returns>
    public HashSet<string> GetRequiredWorkflowEventDefNames()
    {
        return _requiredWorkflowEventDefNames;
    }
    
    internal void AddWorkflowEventDefName(string name) 
    {
        _requiredWorkflowEventDefNames.Add(name);
    }
    
    internal void AddWorkflowEventDefToRegister(ThrowEventNodeOutput node)
    {
        _workflowEventsToRegister.Add(node);
    }
    
    internal ThreadRetentionPolicy? GetDefaultThreadRetentionPolicy() 
    {
        return _defaultThreadRetentionPolicy;
    }
    
    /// <summary>
    /// Sets the retention policy for all ThreadRun's belong to this WfSpec.
    /// 
    /// Note that each Thread can override the configured Retention Policy by
    /// using WorkflowThread#withRetentionPolicy.
    /// </summary>
    /// <param name="policy">
    /// It is the Workflow Retention Policy.
    /// </param>
    /// <returns>
    /// This Workflow.
    /// </returns>
    public Workflow WithDefaultThreadRetentionPolicy(ThreadRetentionPolicy? policy)
    {
        _defaultThreadRetentionPolicy = policy;
        
        return this;
    }
    
    /// <summary>
    /// Defines the type of update to perform when saving the WfSpec:
    /// AllowedUpdateType.ALL (Default): Creates a new WfSpec with a different version (either major or revision).
    /// AllowedUpdateType.MINOR_REVISION_ONLY: Creates a new WfSpec with a different revision if the change is a major version it fails.
    /// AllowedUpdateType.NONE: Fail with the ALREADY_EXISTS response code.
    /// </summary>
    /// <param name="allowedUpdateType">
    /// It is the type of allowed update.
    /// </param>
    /// <returns>
    /// This Workflow.
    /// </returns>
    public Workflow WithUpdateType(AllowedUpdateType allowedUpdateType) 
    {
        _spec.AllowedUpdates = allowedUpdateType;
        
        return this;
    }
    
    /// <summary>
    /// Sets the retention policy for all WfRun's created by this WfSpec.
    /// 
    /// using WorkflowThread#withRetentionPolicy.
    /// </summary>
    /// <param name="policy">
    /// It is the Workflow Retention Policy.
    /// </param>
    /// <returns>
    /// This Workflow.
    /// </returns>
    public Workflow WithRetentionPolicy(WorkflowRetentionPolicy policy) 
    {
        _wfRetentionPolicy = policy;
        
        return this;
    }
    
    /// <summary>
    /// Checks if the WfSpec exists for a given version
    /// </summary>
    /// <param name="client">
    /// It is an LHClient.
    /// </param>
    /// <param name="majorVersion">
    /// This is an optional parameter which represents the workflow Major Version.
    /// </param>
    /// <returns>
    /// True if the workflow spec is registered with/without Major Version or false otherwise
    /// </returns>
    public bool DoesWfSpecExist(LittleHorseClient client, int? majorVersion = null) 
    {
        try
        {
            // TODO: LH-282, support revision versioning here.
            var wfSpecId = new WfSpecId { Name = _name };
            if (majorVersion != null)
            {
                wfSpecId.MajorVersion = majorVersion.Value;
            }
            
            client.GetWfSpec(wfSpecId);
            
            return true;
        }
        catch (RpcException ex) 
        {
            if (ex.StatusCode == StatusCode.NotFound)
            {
                return false;
            }

            throw;
        }
    }
    
    /// <summary>
    /// Returns the associated PutWfSpecRequest in JSON form.
    /// </summary>
    /// <returns>
    /// The associated PutWfSpecRequest in JSON form.
    /// </returns>
    public string? CompileWfToJson() 
    {
        try 
        {
            PutWfSpecRequest wfSpec = Compile();
            return LHMappingHelper.ProtoToJson(wfSpec);
        } 
        catch (InvalidProtocolBufferException exn) 
        {
            throw new Exception($"Cannot compile wfSpec: {exn.Message}");
        }
    }
    
    /// <summary>
    /// Writes out the PutWfSpecRequest in JSON form in a directory.
    /// </summary>
    /// <param name="directory">
    /// It is the location to save the resources.
    /// </param>
    public void CompileAndSaveToDisk(string directory) 
    {
        PutWfSpecRequest wf = Compile();
        string wfFileName = wf.Name + LHConstants.SuffixCompiledWfFileName;
        try
        {
            _logger!.LogInformation($"Saving WfSpec to {wfFileName}");
            SaveProtoToFile(directory, wfFileName, wf);
        }
        catch (Exception e)
        {
            throw new Exception($"Something occurred trying to save file {wfFileName} to disk", e);
        }
    }

    private void SaveProtoToFile(string directory, string fileName, IMessage content) 
    {
        var parentPath = Path.GetFullPath(Path.Combine(AppContext.BaseDirectory, "../../.."));
        var directoryPath = Path.Combine(parentPath, directory);
        Directory.CreateDirectory(directoryPath);
        string filePath = Path.Combine(directoryPath, fileName);
        string? json = LHMappingHelper.ProtoToJson(content);
        
        File.WriteAllText(filePath, json);
    }
}